File: pc/catalog.py
"""Local podcast catalog — no third-party metadata proxy.

Search: Apple iTunes Search API
Trending: 中文播客榜 xyzrank + Apple Top Podcasts RSS-JSON
Episodes: the show's public RSS (parsed locally)
"""

from __future__ import annotations

import re
import time
from typing import Any

from core import Episode, Show, _get_json, resolve_source, search_shows

XYZRANK_PODCASTS = "https://xyzrank.com/api/podcasts"
APPLE_TOP = "https://itunes.apple.com/{country}/rss/toppodcasts/limit={limit}/json"

_TREND_TTL = 30 * 60
_trend_cache: dict[str, tuple[float, tuple[str, list[Show]]]] = {}


def _apple_id_from_url(url: str) -> str:
    m = re.search(r"id(\d+)", url or "")
    return m.group(1) if m else ""


def _link_map(links: Any) -> dict[str, str]:
    out: dict[str, str] = {}
    if not isinstance(links, list):
        return out
    for item in links:
        if not isinstance(item, dict):
            continue
        name = str(item.get("name") or "").strip().lower()
        url = str(item.get("url") or "").strip()
        if name and url:
            out[name] = url
    return out


def _show_from_xyzrank(item: dict[str, Any]) -> Show | None:
    links = _link_map(item.get("links"))
    apple_url = links.get("apple", "")
    apple_id = _apple_id_from_url(apple_url)
    rss = links.get("rss", "").strip()
    xyz = (links.get("xyz") or links.get("website") or "").strip()
    name = str(item.get("name") or "").strip()
    if not name:
        return None
    count = item.get("trackCount") or 0
    try:
        count = int(count)
    except (TypeError, ValueError):
        count = 0
    rank = item.get("rank") or 0
    try:
        rank = int(rank)
    except (TypeError, ValueError):
        rank = 0
    return Show(
        id=apple_id or str(item.get("id") or ""),
        name=name,
        author=str(item.get("authorsText") or ""),
        artwork=str(item.get("logoURL") or ""),
        feed_url=rss or xyz,
        episode_count=count,
        country="CN",
        rank=rank,
    )


def _show_from_apple_entry(entry: dict[str, Any], rank: int, country: str) -> Show | None:
    attrs = (entry.get("id") or {}).get("attributes") or {}
    apple_id = str(attrs.get("im:id") or "")
    name = str((entry.get("im:name") or {}).get("label") or "").strip()
    if not name and not apple_id:
        return None
    images = entry.get("im:image") or []
    artwork = ""
    if isinstance(images, list) and images:
        last = images[-1]
        artwork = last.get("label") if isinstance(last, dict) else ""
    elif isinstance(images, dict):
        artwork = str(images.get("label") or "")
    page = str((entry.get("id") or {}).get("label") or "")
    if not apple_id:
        apple_id = _apple_id_from_url(page)
    return Show(
        id=apple_id,
        name=name or "Untitled",
        author=str((entry.get("im:artist") or {}).get("label") or ""),
        artwork=str(artwork or ""),
        feed_url="",
        episode_count=0,
        country=country.upper(),
        rank=rank,
    )


async def search(query: str, limit: int = 20, country: str = "CN") -> list[Show]:
    """Search Apple's podcast directory. Tries the requested store, then US."""
    q = (query or "").strip()
    if not q:
        return []
    country = (country or "CN").upper()
    shows = await search_shows(q, country=country, limit=limit)
    if shows or country == "US":
        return shows
    extra = await search_shows(q, country="US", limit=limit)
    return extra


async def fetch_xyzrank(limit: int = 40) -> list[Show]:
    data = await _get_json(XYZRANK_PODCASTS, params={"limit": max(1, min(int(limit), 100)), "offset": 0})
    raw = data.get("items") if isinstance(data, dict) else None
    if not isinstance(raw, list):
        nested = data.get("data") if isinstance(data, dict) else None
        if isinstance(nested, dict):
            raw = nested.get("podcasts") or nested.get("items")
        elif isinstance(nested, list):
            raw = nested
    if not isinstance(raw, list):
        return []
    shows: list[Show] = []
    for item in raw:
        if not isinstance(item, dict):
            continue
        show = _show_from_xyzrank(item)
        if show and (show.feed_url or show.id):
            shows.append(show)
    return shows


async def fetch_apple_charts(country: str = "US", limit: int = 40) -> list[Show]:
    cc = (country or "US").lower()
    lim = max(1, min(int(limit), 200))
    url = APPLE_TOP.format(country=cc, limit=lim)
    data = await _get_json(url)
    entries = ((data.get("feed") or {}).get("entry")) if isinstance(data, dict) else None
    if entries is None:
        return []
    if isinstance(entries, dict):
        entries = [entries]
    shows: list[Show] = []
    for i, entry in enumerate(entries or []):
        if not isinstance(entry, dict):
            continue
        show = _show_from_apple_entry(entry, i + 1, cc)
        if show:
            shows.append(show)
    return shows


async def _trending_uncached(source: str) -> tuple[str, list[Show]]:
    src = (source or "cn").lower()
    if src in ("apple", "intl", "international", "us"):
        return "apple", await fetch_apple_charts("US", 40)

    try:
        shows = await fetch_xyzrank(40)
        if shows:
            return "xyzrank", shows
    except Exception:
        shows = []

    # xyzrank down or empty → Apple China chart so the 中文 tab still works
    fallback = await fetch_apple_charts("CN", 40)
    return "apple-cn", fallback


async def trending(source: str = "cn") -> tuple[str, list[Show]]:
    """
    source:
      - cn / xyzrank → 中文播客榜
      - apple / intl → Apple Top (US)
    """
    src = (source or "cn").lower()
    key = "apple" if src in ("apple", "intl", "international", "us") else "cn"
    now = time.monotonic()
    hit = _trend_cache.get(key)
    if hit and now - hit[0] < _TREND_TTL:
        return hit[1]
    try:
        result = await _trending_uncached(key)
        _trend_cache[key] = (now, result)
        return result
    except Exception:
        if hit:
            return hit[1]
        raise


async def list_episodes(src: str, country: str = "CN") -> tuple[Show, list[Episode]]:
    """Load a show and every episode in its public RSS."""
    return await resolve_source(src, country=country)

File: android/app/src/main/java/com/omnix/podstash/data/Catalog.kt
package com.omnix.podstash.data

import org.json.JSONArray
import org.json.JSONObject
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserFactory
import java.io.StringReader
import java.net.URLEncoder
import java.util.Locale

object Catalog {
    private const val ITUNES_SEARCH = "https://itunes.apple.com/search"
    private const val ITUNES_LOOKUP = "https://itunes.apple.com/lookup"
    private const val XYZRANK = "https://xyzrank.com/api/podcasts"
    private const val APPLE_TOP = "https://itunes.apple.com/%s/rss/toppodcasts/limit=%d/json"

    fun search(query: String, country: String = "CN", limit: Int = 20): List<Show> {
        val q = query.trim()
        if (q.isEmpty()) return emptyList()
        val cc = country.uppercase(Locale.ROOT)
        var shows = searchItunes(q, cc, limit)
        if (shows.isEmpty() && cc != "US") shows = searchItunes(q, "US", limit)
        return shows
    }

    fun trending(source: String): Pair<String, List<Show>> {
        val src = source.lowercase(Locale.ROOT)
        if (src in listOf("apple", "intl", "international", "us")) {
            return "apple" to appleCharts("US", 40)
        }
        return try {
            val xyz = xyzrank(40)
            if (xyz.isNotEmpty()) "xyzrank" to xyz else "apple-cn" to appleCharts("CN", 40)
        } catch (_: Exception) {
            "apple-cn" to appleCharts("CN", 40)
        }
    }

    fun resolve(src: String, country: String = "CN"): Pair<Show, List<Episode>> {
        val s = src.trim()
        require(s.isNotEmpty()) { "请输入节目 ID、链接或 RSS" }
        val lower = s.lowercase(Locale.ROOT)
        when {
            "youzhiyouxing.cn/materials/" in lower -> error("有知有行 materials 请在 PC 端打开")
            "xiaoyuzhoufm.com/podcast/" in lower || "xiaoyuzhoufm.com/episode/" in lower ->
                return resolveXiaoyuzhou(s, country)
            "ximalaya.com/album/" in lower -> {
                val id = Regex("""album/(\d+)""").find(s)?.groupValues?.get(1)
                    ?: error("不是有效的喜马拉雅专辑链接")
                return fetchRss(Show(id = id, name = "", feedUrl = "https://www.ximalaya.com/album/$id.xml"))
            }
            s.all { it.isDigit() } -> return fetchRss(lookup(s) ?: error("找不到 Apple Podcast ID: $s"))
            "podcasts.apple.com" in lower -> {
                val id = Regex("""id(\d+)""").find(s)?.groupValues?.get(1)
                    ?: error("无法从 Apple 链接解析 ID")
                return fetchRss(lookup(id) ?: error("找不到 Apple Podcast ID: $id"))
            }
            lower.startsWith("http://") || lower.startsWith("https://") ->
                return fetchRss(Show(id = "", name = "", feedUrl = s))
            else -> error("无法识别来源")
        }
    }

    private fun searchItunes(term: String, country: String, limit: Int): List<Show> {
        val url =
            "$ITUNES_SEARCH?media=podcast&entity=podcast&country=$country&limit=$limit&term=" +
                URLEncoder.encode(term, "UTF-8")
        val arr = Http.getJson(url).optJSONArray("results") ?: JSONArray()
        val out = mutableListOf<Show>()
        for (i in 0 until arr.length()) {
            val o = arr.optJSONObject(i) ?: continue
            val feed = o.optString("feedUrl").trim()
            if (feed.isEmpty()) continue
            out += Show(
                id = o.opt("collectionId")?.toString() ?: "",
                name = o.optString("collectionName").ifBlank { o.optString("trackName") },
                author = o.optString("artistName"),
                artwork = o.optString("artworkUrl600").ifBlank { o.optString("artworkUrl100") },
                feedUrl = feed,
                episodeCount = o.optInt("trackCount"),
                country = o.optString("country").ifBlank { country },
            )
        }
        return out
    }

    fun lookup(appleId: String): Show? {
        val data = Http.getJson("$ITUNES_LOOKUP?id=$appleId&entity=podcast")
        val arr = data.optJSONArray("results") ?: return null
        for (i in 0 until arr.length()) {
            val o = arr.optJSONObject(i) ?: continue
            val feed = o.optString("feedUrl").trim()
            if (feed.isEmpty()) continue
            return Show(
                id = o.opt("collectionId")?.toString() ?: appleId,
                name = o.optString("collectionName").ifBlank { o.optString("trackName") },
                author = o.optString("artistName"),
                artwork = o.optString("artworkUrl600").ifBlank { o.optString("artworkUrl100") },
                feedUrl = feed,
                episodeCount = o.optInt("trackCount"),
                country = o.optString("country"),
            )
        }
        return null
    }

    private fun xyzrank(limit: Int): List<Show> {
        val data = Http.getJson("$XYZRANK?limit=$limit&offset=0")
        val items = data.optJSONArray("items") ?: JSONArray()
        val out = mutableListOf<Show>()
        for (i in 0 until items.length()) {
            val o = items.optJSONObject(i) ?: continue
            val links = linkMap(o.optJSONArray("links"))
            val appleUrl = links["apple"].orEmpty()
            val appleId = Regex("""id(\d+)""").find(appleUrl)?.groupValues?.get(1).orEmpty()
            val rss = links["rss"].orEmpty()
            val xyz = links["xyz"] ?: links["website"].orEmpty()
            val name = o.optString("name").trim()
            if (name.isEmpty()) continue
            val feed = rss.ifBlank { xyz }
            if (feed.isEmpty() && appleId.isEmpty()) continue
            out += Show(
                id = appleId.ifBlank { o.optString("id") },
                name = name,
                author = o.optString("authorsText"),
                artwork = o.optString("logoURL"),
                feedUrl = feed,
                episodeCount = o.optInt("trackCount"),
                country = "CN",
                rank = o.optInt("rank"),
            )
        }
        return out
    }

    private fun appleCharts(country: String, limit: Int): List<Show> {
        val data = Http.getJson(APPLE_TOP.format(country.lowercase(Locale.ROOT), limit.coerceIn(1, 200)))
        val feed = data.optJSONObject("feed") ?: return emptyList()
        val raw = feed.opt("entry") ?: return emptyList()
        val entries: JSONArray = when (raw) {
            is JSONArray -> raw
            is JSONObject -> JSONArray().put(raw)
            else -> return emptyList()
        }
        val out = mutableListOf<Show>()
        for (i in 0 until entries.length()) {
            val e = entries.optJSONObject(i) ?: continue
            val idObj = e.optJSONObject("id")
            val appleId = idObj?.optJSONObject("attributes")?.optString("im:id").orEmpty()
            val name = e.optJSONObject("im:name")?.optString("label").orEmpty()
            if (name.isEmpty() && appleId.isEmpty()) continue
            val images = e.optJSONArray("im:image")
            val art = if (images != null && images.length() > 0) {
                images.optJSONObject(images.length() - 1)?.optString("label").orEmpty()
            } else ""
            out += Show(
                id = appleId,
                name = name.ifBlank { "Untitled" },
                author = e.optJSONObject("im:artist")?.optString("label").orEmpty(),
                artwork = art,
                feedUrl = "",
                rank = i + 1,
                country = country.uppercase(Locale.ROOT),
            )
        }
        return out
    }

    private fun resolveXiaoyuzhou(url: String, country: String): Pair<Show, List<Episode>> {
        val html = Http.getString(url)
        val title = xiaoyuzhouTitle(html).ifBlank { error("无法读取小宇宙节目标题") }
        val found = search(title, country, 10)
        val match = found.firstOrNull { it.name.replace("\\s".toRegex(), "").equals(title.replace("\\s".toRegex(), ""), true) }
            ?: found.firstOrNull()
            ?: error("未在 Apple 目录中找到「$title」的 RSS")
        val show = if (match.id.isNotBlank()) lookup(match.id) ?: match else match
        return fetchRss(show)
    }

    private fun xiaoyuzhouTitle(html: String): String {
        val next = Regex("""<script id="__NEXT_DATA__"[^>]*>(.*?)</script>""", RegexOption.DOT_MATCHES_ALL)
            .find(html)?.groupValues?.get(1)
        if (!next.isNullOrBlank()) {
            try {
                val data = JSONObject(next)
                val page = data.optJSONObject("props")?.optJSONObject("pageProps")
                val pod = page?.optJSONObject("podcast")
                    ?: page?.optJSONObject("episode")?.optJSONObject("podcast")
                val t = pod?.optString("title").orEmpty()
                if (t.isNotBlank()) return t
            } catch (_: Exception) {
            }
        }
        return Regex("""property=["']og:title["'][^>]*content=["']([^"']+)""")
            .find(html)?.groupValues?.get(1).orEmpty()
    }

    fun fetchRss(show: Show): Pair<Show, List<Episode>> {
        val feed = show.feedUrl.ifBlank {
            if (show.id.isNotBlank()) lookup(show.id)?.feedUrl.orEmpty() else ""
        }
        require(feed.isNotBlank()) { "该节目没有 RSS" }
        val xml = Http.getString(feed)
        val (title, episodes) = parseRss(xml)
        val merged = show.copy(
            name = if (title.isNotBlank() && title != "Podcast") title else show.name.ifBlank { title },
            feedUrl = feed,
            episodeCount = episodes.size,
        )
        return merged to episodes
    }

    private fun parseRss(xml: String): Pair<String, List<Episode>> {
        val factory = XmlPullParserFactory.newInstance().apply { isNamespaceAware = true }
        val p = factory.newPullParser()
        p.setInput(StringReader(xml))
        var showTitle = "Podcast"
        val episodes = mutableListOf<Episode>()
        var event = p.eventType
        var inItem = false
        var inChannel = false
        var title = ""
        var guid = ""
        var pub = ""
        var duration = ""
        var enclosure = ""
        var size = 0L
        fun flushItem() {
            if (enclosure.isNotBlank()) {
                episodes += Episode(
                    index = episodes.size,
                    title = title.ifBlank { "Untitled" },
                    audioUrl = enclosure,
                    published = pub.take(10),
                    duration = duration,
                    guid = guid.ifBlank { enclosure },
                    size = size,
                )
            }
            title = ""; guid = ""; pub = ""; duration = ""; enclosure = ""; size = 0L
        }
        while (event != XmlPullParser.END_DOCUMENT) {
            val name = p.name?.substringAfterLast(':') ?: ""
            when (event) {
                XmlPullParser.START_TAG -> when (name.lowercase(Locale.ROOT)) {
                    "channel", "feed" -> inChannel = true
                    "item", "entry" -> {
                        inItem = true
                        title = ""; guid = ""; pub = ""; duration = ""; enclosure = ""; size = 0L
                    }
                    "title" -> {
                        val t = nextText(p)
                        if (inItem) title = t else if (inChannel && showTitle == "Podcast") showTitle = t.ifBlank { showTitle }
                    }
                    "guid", "id" -> if (inItem) guid = nextText(p)
                    "pubdate", "published", "updated" -> if (inItem) pub = nextText(p)
                    "duration" -> if (inItem) duration = nextText(p)
                    "enclosure" -> if (inItem) {
                        val url = p.getAttributeValue(null, "url").orEmpty()
                        val typ = p.getAttributeValue(null, "type").orEmpty().lowercase(Locale.ROOT)
                        if (url.isNotBlank() && (typ.startsWith("audio") || typ.startsWith("video") ||
                                url.contains(Regex("""\.(mp3|m4a|aac|ogg|opus|wav|flac)(\?|$)""", RegexOption.IGNORE_CASE)))
                        ) {
                            enclosure = url
                            size = p.getAttributeValue(null, "length")?.toLongOrNull() ?: 0L
                        }
                    }
                    "link" -> if (inItem && enclosure.isEmpty()) {
                        val href = p.getAttributeValue(null, "href").orEmpty()
                        val rel = p.getAttributeValue(null, "rel").orEmpty()
                        val typ = p.getAttributeValue(null, "type").orEmpty()
                        if (href.isNotBlank() && (rel == "enclosure" || typ.startsWith("audio"))) enclosure = href
                    }
                }
                XmlPullParser.END_TAG -> when (name.lowercase(Locale.ROOT)) {
                    "item", "entry" -> {
                        flushItem()
                        inItem = false
                    }
                }
            }
            event = p.next()
        }
        return showTitle to episodes
    }

    private fun nextText(p: XmlPullParser): String = try {
        p.nextText()?.trim().orEmpty()
    } catch (_: Exception) {
        ""
    }

    private fun linkMap(arr: JSONArray?): Map<String, String> {
        if (arr == null) return emptyMap()
        val out = mutableMapOf<String, String>()
        for (i in 0 until arr.length()) {
            val o = arr.optJSONObject(i) ?: continue
            val n = o.optString("name").trim().lowercase(Locale.ROOT)
            val u = o.optString("url").trim()
            if (n.isNotBlank() && u.isNotBlank()) out[n] = u
        }
        return out
    }
}
